src/arithmetic.jl

```diff
@@ -1,17 +1,34 @@
+### Mag
 for (jf, af) in [(:+, :add!), (:-, :sub!), (:*, :mul!), (:/, :div!)]
-    @eval function Base.$jf(x::T, y::T) where {T<:MagOrRef}
-        z = T()
-        $af(z, x, y)
-        z
-    end
+    @eval Base.$jf(x::MagOrRef, y::MagOrRef) = $af(zero(x), x, y)
 end
+Base.:+(x::MagOrRef, y::Integer) = add!(zero(x), x, convert(UInt, y))
+Base.:+(x::Integer, y::MagOrRef) = add!(zero(y), y, convert(UInt, x))
+Base.:*(x::MagOrRef, y::Integer) = mul!(zero(x), x, convert(UInt, y))
+Base.:*(x::Integer, y::MagOrRef) = mul!(zero(y), y, convert(UInt, x))
+Base.:/(x::MagOrRef, y::Integer) = div!(zero(x), x, convert(UInt, y))
 
-function Base.:(^)(x::T, k::Integer) where {T<:MagOrRef}
-    z = T()
-    pow!(z, x, convert(UInt, k))
-    z
-end
+Base.inv(x::MagOrRef) = inv!(zero(x), x)
+Base.:(^)(x::MagOrRef, e::Integer) = pow!(zero(x), x, convert(UInt, e))
+Base.sqrt(x::MagOrRef) = sqrt!(zero(x), x)
+rsqrt(x::MagOrRef) = rsqrt!(zero(x), x)
+Base.hypot(x::MagOrRef, y::MagOrRef) = hypot!(zero(x), x, y)
+root(x::MagOrRef, n::Integer) = root!(zero(x), x, convert(UInt, n))
+Base.log(x::MagOrRef) = log!(zero(x), x)
+neglog(x::MagOrRef) = neg_log!(zero(x), x)
+Base.log1p(x::MagOrRef) = log1p!(zero(x), x)
+Base.exp(x::MagOrRef) = exp!(zero(x), x)
+expinv(x::MagOrRef) = expinv!(zero(x), x)
+Base.expm1(x::MagOrRef) = expm1!(zero(x), x)
+Base.atan(x::MagOrRef) = atan!(zero(x), x)
+Base.cosh(x::MagOrRef) = cosh!(zero(x), x)
+Base.sinh(x::MagOrRef) = sinh!(zero(x), x)
+
+Base.min(x::MagOrRef, y::MagOrRef) = min!(zero(x), x, y)
+Base.max(x::MagOrRef, y::MagOrRef) = max!(zero(x), x, y)
+Base.minmax(x::MagOrRef, y::MagOrRef) = (min(x, y), max(x, y))
```

Comment:
do we want to write those functions like this of through `@eval` in a loop (with 5 exceptions?)

Reply:
Maybe a loop will look nicer, strictly more lines but easier to change. Will fix that!

Reply:
Actually it fit on three lines, so much nicer!